### PR TITLE
Use docker image repo in ClusterSpec

### PR DESCRIPTION
**Pull request by hmusum**

### CODE REVIEW

File: config-provisioning/src/main/java/com/yahoo/config/provision/HostSpec.java [hunk 1]
```diff
@@ -77,6 +88,7 @@ public HostSpec(String hostname, List<String> aliases, Optional<Flavor> flavor,
         this.version = Objects.requireNonNull(version, "Version cannot be null but can be empty");;
         this.networkPorts = Objects.requireNonNull(networkPorts, "Network ports cannot be null but can be empty");;
         this.requestedResources = Objects.requireNonNull(requestedResources, "RequestedResources cannot be null");
+        this.dockerImageRepo = Objects.requireNonNull(dockerImageRepo, "Version cannot be null but can be empty");;
```
Comment (freva) [inline config-provisioning/src/main/java/com/yahoo/config/provision/HostSpec.java]:
```suggestion
        this.dockerImageRepo = Objects.requireNonNull(dockerImageRepo, "Docker image repo cannot be null but can be empty");
```

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java [hunk 2]
```diff
@@ -88,7 +88,7 @@ public void prepare() {
                 Version targetVersion = infrastructureVersions.getTargetVersionFor(nodeType);
                 hostSpecs = provisioner.prepare(
                         application.getApplicationId(),
-                        application.getClusterSpecWithVersion(targetVersion),
+                        application.getClusterSpecWithVersion(targetVersion, Optional.empty()), // TODO: Use dockerImageRepo from zk
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java]: Does this apply here? We are potentially allocating new nodes here, which do not have any docker image associated with them. Unless you mean the docker image from https://github.com/vespa-engine/vespa/blob/master/node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/DockerImages.java
?

File: config-provisioning/src/test/java/com/yahoo/config/provision/serialization/AllocatedHostsSerializerTest.java [hunk 3]
```diff
@@ -72,6 +73,7 @@ private void assertAllocatedHosts(AllocatedHosts expectedHosts, NodeFlavors conf
             assertEquals(expectedHost.networkPorts(), deserializedHost.networkPorts());
             assertEquals(expectedHost.aliases(), deserializedHost.aliases());
             assertEquals(expectedHost.requestedResources(), deserializedHost.requestedResources());
+            assertEquals(expectedHost.dockerImageRepo().orElse(""), deserializedHost.dockerImageRepo().orElse(""));
```
Comment (freva) [inline config-provisioning/src/test/java/com/yahoo/config/provision/serialization/AllocatedHostsSerializerTest.java]: I think you can just compare the Optionals directly
```suggestion
            assertEquals(expectedHost.dockerImageRepo(), deserializedHost.dockerImageRepo());
```

**Review by freva — APPROVED** [review]
LGTM.

I think the docker image repo should (at some point) become https://github.com/vespa-engine/vespa/blob/master/config-provisioning/src/main/java/com/yahoo/config/provision/DockerImage.java instead of string.

`ClusterSpec` should also just get a builder so we don't need to update 20 tests every time we add a new field. Plus the builder methods could take in things as VespaVersion as both `String` or `Version`, so 
```
ClusterSpec.request(ClusterSpec.Type.content, ClusterSpec.Id.from("test"),
Version.fromString("6.42"), false, Optional.empty(), Optional.empty())
```
Would be something like
```
ClusterSpec.builder(ClusterSpec.Type.content, "test").version("6.42").build()
```

**Comment by hmusum** [comment]
Yeah, I agree that we should have a builder for ClusterSpec, will look into it (in another PR).

Regarding DockerImage: Yes, I thought about using that now, but decided to wait. I'll look at that in a forthcoming PR.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java (hunk 2 above)
Comment (hmusum) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java]: No, probably not. I'll just remove it for now.

**Review by freva — APPROVED** [review]

**Pull request merged**